Universal API command

tzdybal opened the pull request:
New CLI command (`api`) that allows executing any API method from any module.
Result is marshaled as pretty-printed JSON.

This resolves #353.

whilei commented on cmd/geth/apicmd.go:
```diff
@@ -0,0 +1,113 @@
+		return res.Result, nil
+	}
+
+	return nil, errors.New("No API response")
```
If `res.Result` is nil, it probably means we actually got an error from the API -- maybe we can use `rpc.JSONErrResponse` and print the error message

whilei commented on cmd/geth/apicmd.go:
```diff
@@ -0,0 +1,113 @@
+		"id":      new(int64),
+		"method":  module + "_" + method,
+		"jsonrpc": "2.0",
+		"params":  []interface{}{args},
```
We may need to parse `args` here to their probable/matching types, since all `ctx.Args()` (I think..?) will be `string`s... 

and may need to establish a syntax for passing "object" values, like

```json
{"jsonrpc":"2.0","method":"eth_sendTransaction","params":[{"from": "0x396599f365093186742c17aab158bf515e978bc7", "gas": "0x5208", "gasPrice": "0x02540be400", "to": "0xa02cee0fc1d3fb4dde86b79fe93e4140671fd949"}],"id":1}'
```

, and even arrays containing objects _and_ other values, like blockNumber (see `eth_traceCall` for an example). 

... Maybe the simplest way is to require `ctx.Args()[2:]` to be a single JSON string instead?

whilei commented:
```shell
~  ⟠ geth --chain morden api eth getBlock latest
No API response
~  ⟠ geth --chain morden api eth getBlock 0
No API response
~  ⟠ geth --chain morden api eth blockNumber
"0x1c3069"
~  ⟠ geth --chain morden api eth getBlock "latest"
No API response
~  ⟠ geth --chain morden api eth getBlock '"latest"'
No API response
~  ⟠ geth --chain morden api eth getBlock "0"
No API response
~  ⟠ geth --chain morden api eth getBlock "['0']"
No API response
~  ⟠ geth --chain morden api eth getBlock "[0]"
No API response
```

whilei commented on cmd/geth/apicmd.go:
```diff
@@ -0,0 +1,113 @@
+func callRPC(ctx *cli.Context, client rpc.Client) (interface{}, error) {
+	var (
+		module = ctx.Args()[0]
+		method = ctx.Args()[1]
```
Should add check, perhaps in `verifyArguments` for existence of required args. For example:

```shell
~  ⟠ geth --chain morden api
runtime error: index out of range
```

tzdybal replied:
Fixed.

tzdybal commented:
I addressed all issues. API method arguments should be passed space separated, with correct quotation, so they are correct JSON values. This way the simple cases are still simple (no need to wrap arguments into JSON code), and complex cases are also possible.
Here are some example invocations:
`geth --chain morden api eth getBlockByNumber 1 true`
`geth --chain morden api eth getBlockByNumber '"latest"' true`
`geth --chain morden api eth sendTransaction '{"from": "0x396599f365093186742c17aab158bf515e978bc7", "gas": "0x5208", "gasPrice": "0x02540be400", "to": "0xa02cee0fc1d3fb4dde86b79fe93e4140671fd949"}'`

whilei commented on cmd/geth/apicmd.go:
```diff
@@ -0,0 +1,124 @@
+		Usage:           "Run any API command",
+		SkipFlagParsing: true,
+		Description: `
+The api command allows to run any method defined in any API module.
```
Let's add something about what to expect as result... like 

> Output will be in JSON format.

whilei commented:
Will also resolve #355 ✅

whilei commented on cmd/geth/apicmd.go:
```diff
@@ -0,0 +1,137 @@
+		Action:          execAPI,
+		Name:            "api",
+		Usage:           "Run any API command",
+		SkipFlagParsing: true,
```
Just one final issue I found; skipping flag parsing disabled `--help` usage -- this can be fixed by just removing the line.

```shell
~/gocode/src/github.com/ethereumproject/go-ethereum tzdybal-api_command= ⟠ geth api --help
dial unix /Users/ia/Library/EthereumClassic/mainnet/geth.ipc: connect: no such file or directory

# for comparison
~/gocode/src/github.com/ethereumproject/go-ethereum tzdybal-api_command= ⟠ geth monitor --help
monitor [command options] [arguments...]

The Geth monitor is a tool to collect and visualize various internal metrics
gathered by the node, supporting different chart types as well as the capacity
to display multiple metrics simultaneously.


OPTIONS:
	--attach value	API endpoint to attach to (default: "ipc:/Users/ia/Library/EthereumClassic/mainnet/geth.ipc")
	--rows value	Maximum rows in the chart grid (default: 5)
	--refresh value	Refresh interval in seconds (default: 3)
```
